ClassNotFoundException in native mode with custom IdentifierGenerator after upgrading to Hibernate 7

### Describe the bug

After upgrading to Hibernate ORM 7, our custom IdentifierGenerator no longer works in native mode unless we explicitly annotate it with @RegisterForReflection. This was not required in previous versions.

### Expected behavior

Custom generators should either:

- Be automatically registered for reflection (as in Hibernate 6)
- Or the Migration guide should mention this requirement in context of update Hibernate to 7.0

### Actual behavior

``` 
2025-06-26 14:47:24,744 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: jakarta.persistence.PersistenceException: [PersistenceUnit: <default>] Unable to build Hibernate SessionFactory
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:127)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:94)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:73)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:90)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:66)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:163)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:63)
        at java.base@21.0.7/java.lang.Thread.runWith(Thread.java:1596)
        at java.base@21.0.7/java.lang.Thread.run(Thread.java:1583)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:896)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:872)
Caused by: org.hibernate.boot.registry.classloading.spi.ClassLoadingException: Unable to load class [org.example.generator.MyCustomIdGenerator]
        at io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService.classForName(FlatClassLoaderService.java:38)
        at org.hibernate.boot.model.internal.GeneratorStrategies.generatorClass(GeneratorStrategies.java:107)
        at org.hibernate.boot.model.internal.GeneratorBinder.lambda$instantiateNamedStrategyGenerator$8(GeneratorBinder.java:781)
        at org.hibernate.mapping.SimpleValue.createGenerator(SimpleValue.java:421)
        at org.hibernate.tuple.entity.EntityMetamodel.buildIdGenerator(EntityMetamodel.java:514)
        at org.hibernate.tuple.entity.EntityMetamodel.lambda$new$0(EntityMetamodel.java:154)
        at org.hibernate.tuple.entity.EntityMetamodel.<init>(EntityMetamodel.java:176)
        at org.hibernate.tuple.entity.EntityMetamodel.<init>(EntityMetamodel.java:153)
        at org.hibernate.metamodel.spi.RuntimeModelCreationContext.createEntityMetamodel(RuntimeModelCreationContext.java:75)
        at org.hibernate.persister.entity.AbstractEntityPersister.<init>(AbstractEntityPersister.java:497)
        at org.hibernate.persister.entity.SingleTableEntityPersister.<init>(SingleTableEntityPersister.java:116)
        at java.base@21.0.7/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at java.base@21.0.7/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
        at org.hibernate.persister.internal.PersisterFactoryImpl.createEntityPersister(PersisterFactoryImpl.java:70)
        at org.hibernate.metamodel.model.domain.internal.MappingMetamodelImpl.processBootEntities(MappingMetamodelImpl.java:217)
        at org.hibernate.metamodel.model.domain.internal.MappingMetamodelImpl.finishInitialization(MappingMetamodelImpl.java:156)
        at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:310)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:92)
        ... 9 more
Caused by: java.lang.ClassNotFoundException: org.example.generator.MyCustomIdGenerator
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:122)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:86)
        at java.base@21.0.7/java.lang.Class.forName(DynamicHub.java:1359)
        at java.base@21.0.7/java.lang.Class.forName(DynamicHub.java:1348)
        at io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService.classForName(FlatClassLoaderService.java:36)
        ... 26 more
```

### How to Reproduce?

1. `git clone https://github.com/gtroitsk/hibernate7-custom-genid-reproducer`
2. `cd hibernate7-custom-genid-reproducer`
3. `mvn clean verify  -Dnative`

### Output of `uname -a` or `ver`

Linux fedora 6.13.9-200.fc41.x86_64

### Output of `java -version`

OpenJDK 64-Bit Server VM Temurin-21.0.7+6 (build 21.0.7+6-LTS, mixed mode, sharing)

### Mandrel or GraalVM version (if different from Java)

MANDREL 23.1.7.0 JDK 21.0.7+6-LTS

### Quarkus version or git rev

999-SNAPSHOT; 3.24.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (Red Hat 3.9.6-7)

### Additional information

_No response_

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/ClassNames.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/ClassNames.java
@@ -62,6 +62,10 @@ public final class ClassNames {
     public static final DotName HIBERNATE_CACHE = createConstant("org.hibernate.Cache");
     public static final DotName PERSISTENCE_UNIT_UTIL = createConstant("jakarta.persistence.PersistenceUnitUtil");
 
+    public static final DotName GENERIC_GENERATOR = createConstant("org.hibernate.annotations.GenericGenerator");
+    public static final DotName ID_GENERATOR_TYPE = createConstant("org.hibernate.annotations.IdGeneratorType");
+    public static final DotName VALUE_GENERATION_TYPE = createConstant("org.hibernate.annotations.ValueGenerationType");
+
     public static final DotName INTERCEPTOR = createConstant("org.hibernate.Interceptor");
     public static final DotName STATEMENT_INSPECTOR = createConstant("org.hibernate.resource.jdbc.spi.StatementInspector");
     public static final DotName FORMAT_MAPPER = createConstant("org.hibernate.type.format.FormatMapper");

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/GraalVMFeatures.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/GraalVMFeatures.java
@@ -26,6 +26,8 @@ public class GraalVMFeatures {
 
     // TODO try to limit registration to those that are actually needed, based on configuration + mapping.
     //   https://github.com/quarkusio/quarkus/pull/32433#issuecomment-1497615958
+    //   See also io.quarkus.hibernate.orm.deployment.JpaJandexScavenger.enlistClassReferences for
+    //   the beginning of a solution (which only handles custom types, not references by name such as 'sequence').
     @BuildStep
     ReflectiveClassBuildItem registerGeneratorAndOptimizerClassesForReflections() {
         return ReflectiveClassBuildItem

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
@@ -94,6 +94,10 @@ public final class JpaJandexScavenger {
             enlistPotentialCdiBeanClasses(collector, annotation);
         }
 
+        enlistPotentialClassReferences(collector, ClassNames.GENERIC_GENERATOR, "type", "strategy");
+        enlistPotentialClassReferences(collector, ClassNames.ID_GENERATOR_TYPE, "value");
+        enlistPotentialClassReferences(collector, ClassNames.VALUE_GENERATION_TYPE, "generatedBy");
+
         for (JpaModelPersistenceUnitContributionBuildItem persistenceUnitContribution : persistenceUnitContributions) {
             enlistExplicitMappings(collector, persistenceUnitContribution);
         }
@@ -214,10 +218,16 @@ public final class JpaJandexScavenger {
             String nodeName) {
         String name = safeGetClassName(packagePrefix, managed, nodeName);
         enlistExplicitClass(collector, name);
-        if (managed instanceof JaxbEntity) {
+        if (managed instanceof JaxbEntity entity) {
             // The call to 'enlistExplicitClass' above may not
             // detect that this class is an entity if it is not annotated
             collector.entityTypes.add(name);
+
+            // Generators may be instantiated reflectively
+            if (entity.getGenericGenerator() != null) {
+                var generator = entity.getGenericGenerator();
+                enlistPotentialClassReference(collector, generator == null ? null : generator.getClazz());
+            }
         }
 
         enlistOrmXmlMappingListeners(collector, packagePrefix, managed.getEntityListenerContainer());
@@ -435,6 +445,47 @@ public final class JpaJandexScavenger {
             DotName beanTypeDotName = beanType.name();
             collector.potentialCdiBeanTypes.add(beanTypeDotName);
         }
+    }
+
+    private void enlistPotentialClassReferences(Collector collector, DotName dotName, String... referenceAttributes) {
+        Collection<AnnotationInstance> jpaAnnotations = index.getAnnotations(dotName);
+
+        if (jpaAnnotations == null) {
+            return;
+        }
+
+        for (AnnotationInstance annotation : jpaAnnotations) {
+            for (String referenceAttribute : referenceAttributes) {
+                var referenceValue = annotation.value(referenceAttribute);
+                if (referenceValue == null) {
+                    continue;
+                }
+                String reference = switch (referenceValue.kind()) {
+                    case CLASS -> referenceValue.asClass().name().toString();
+                    case STRING -> {
+                        String stringRef = referenceValue.asString();
+                        if (stringRef.isEmpty() || index.getClassByName(stringRef) == null) {
+                            // No reference, or reference to a built-in strategy name like 'sequence'
+                            // (which we can't resolve here and handle through GraalVMFeatures.registerGeneratorAndOptimizerClassesForReflections)
+                            yield null;
+                        }
+                        yield stringRef;
+                    }
+                    default -> null;
+                };
+                enlistPotentialClassReference(collector, reference);
+            }
+        }
+    }
+
+    /**
+     * Add the class to the reflective list with only constructor and method access.
+     */
+    private void enlistPotentialClassReference(Collector collector, String reference) {
+        if (reference == null) {
+            return;
+        }
+        collector.javaTypes.add(reference);
     }
 
     /**

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomGenericGeneratorReferencedAsClass.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomGenericGeneratorReferencedAsClass.java
@@ -1,0 +1,15 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+import org.hibernate.annotations.GenericGenerator;
+
+@Entity
+public class EntityWithCustomGenericGeneratorReferencedAsClass {
+    @Id
+    @GeneratedValue(generator = "referenced-as-class")
+    @GenericGenerator(name = "referenced-as-class", type = MyCustomGenericGeneratorReferencedAsClass.class)
+    public String id;
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomGenericGeneratorReferencedAsClassName.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomGenericGeneratorReferencedAsClassName.java
@@ -1,0 +1,15 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+import org.hibernate.annotations.GenericGenerator;
+
+@Entity
+public class EntityWithCustomGenericGeneratorReferencedAsClassName {
+    @Id
+    @GeneratedValue(generator = "referenced-as-class-name")
+    @GenericGenerator(name = "referenced-as-class-name", strategy = "io.quarkus.it.jpa.generatedvalue.MyCustomGenericGeneratorReferencedAsClassName")
+    public String id;
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomIdGeneratorType.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomIdGeneratorType.java
@@ -1,0 +1,13 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+@Entity
+public class EntityWithCustomIdGeneratorType {
+    @Id
+    @GeneratedValue
+    @MyCustomIdGeneratorAnnotation
+    public String id;
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomValueGeneratorType.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/EntityWithCustomValueGeneratorType.java
@@ -1,0 +1,15 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+@Entity
+public class EntityWithCustomValueGeneratorType {
+    @Id
+    @GeneratedValue
+    public Integer id;
+
+    @MyCustomValueGeneratorAnnotation
+    public String customGenerated;
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/GeneratedValueResource.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/GeneratedValueResource.java
@@ -43,6 +43,30 @@ public class GeneratedValueResource {
                 e -> assertThat(e.generated).isNotNull(),
                 e -> assertThat(e.generatedColumn).isNotNull());
 
+        var entity2 = new EntityWithCustomIdGeneratorType();
+        assertThat(entity2.id).isNull();
+        em.persist(entity2);
+        em.flush();
+        assertThat(entity2.id).isEqualTo(MyCustomIdGenerator.STUB_VALUE);
+
+        var entity3 = new EntityWithCustomValueGeneratorType();
+        assertThat(entity3.customGenerated).isNull();
+        em.persist(entity3);
+        em.flush();
+        assertThat(entity3.customGenerated).isEqualTo(MyCustomValueGenerator.STUB_VALUE);
+
+        var entity4 = new EntityWithCustomGenericGeneratorReferencedAsClass();
+        assertThat(entity4.id).isNull();
+        em.persist(entity4);
+        em.flush();
+        assertThat(entity4.id).isEqualTo(MyCustomGenericGeneratorReferencedAsClass.STUB_VALUE);
+
+        var entity5 = new EntityWithCustomGenericGeneratorReferencedAsClassName();
+        assertThat(entity5.id).isNull();
+        em.persist(entity5);
+        em.flush();
+        assertThat(entity5.id).isEqualTo(MyCustomGenericGeneratorReferencedAsClassName.STUB_VALUE);
+
         return "OK";
     }
 }

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomGenericGeneratorReferencedAsClass.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomGenericGeneratorReferencedAsClass.java
@@ -1,0 +1,13 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.id.IdentifierGenerator;
+
+public class MyCustomGenericGeneratorReferencedAsClass implements IdentifierGenerator {
+    public static String STUB_VALUE = MyCustomGenericGeneratorReferencedAsClass.class.getName() + "_STUB_VALUE";
+
+    @Override
+    public Object generate(SharedSessionContractImplementor session, Object object) {
+        return STUB_VALUE;
+    }
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomGenericGeneratorReferencedAsClassName.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomGenericGeneratorReferencedAsClassName.java
@@ -1,0 +1,13 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.id.IdentifierGenerator;
+
+public class MyCustomGenericGeneratorReferencedAsClassName implements IdentifierGenerator {
+    public static String STUB_VALUE = MyCustomGenericGeneratorReferencedAsClassName.class.getName() + "_STUB_VALUE";
+
+    @Override
+    public Object generate(SharedSessionContractImplementor session, Object object) {
+        return STUB_VALUE;
+    }
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomIdGenerator.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomIdGenerator.java
@@ -1,0 +1,13 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.id.IdentifierGenerator;
+
+public class MyCustomIdGenerator implements IdentifierGenerator {
+    public static String STUB_VALUE = MyCustomIdGenerator.class.getName() + "_STUB_VALUE";
+
+    @Override
+    public Object generate(SharedSessionContractImplementor session, Object object) {
+        return STUB_VALUE;
+    }
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomIdGeneratorAnnotation.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomIdGeneratorAnnotation.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.hibernate.annotations.IdGeneratorType;
+
+@IdGeneratorType(MyCustomIdGenerator.class)
+@Retention(RUNTIME)
+@Target({ FIELD, METHOD })
+public @interface MyCustomIdGeneratorAnnotation {
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomValueGenerator.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomValueGenerator.java
@@ -1,0 +1,21 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import java.util.EnumSet;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.generator.BeforeExecutionGenerator;
+import org.hibernate.generator.EventType;
+
+public class MyCustomValueGenerator implements BeforeExecutionGenerator {
+    public static String STUB_VALUE = MyCustomValueGenerator.class.getName() + "_STUB_VALUE";
+
+    @Override
+    public Object generate(SharedSessionContractImplementor session, Object owner, Object currentValue, EventType eventType) {
+        return STUB_VALUE;
+    }
+
+    @Override
+    public EnumSet<EventType> getEventTypes() {
+        return EnumSet.of(EventType.INSERT, EventType.UPDATE);
+    }
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomValueGeneratorAnnotation.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/generatedvalue/MyCustomValueGeneratorAnnotation.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.jpa.generatedvalue;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.hibernate.annotations.ValueGenerationType;
+
+@ValueGenerationType(generatedBy = MyCustomValueGenerator.class)
+@Retention(RUNTIME)
+@Target({ FIELD, METHOD })
+public @interface MyCustomValueGeneratorAnnotation {
+}

